modules/images/image-loading-optimization/optimization.php

```diff
@@ -0,0 +1,174 @@
+		if ( false === $lcp_element || empty( $lcp_element['attributes'] ) ) {
+			// No LCP element at this breakpoint, so nothing to preload.
+			continue;
+		}
+
+		$img_attributes = $lcp_element['attributes'];
+
+		// Prevent preloading src for browsers that don't support imagesrcset on the link element.
+		if ( isset( $img_attributes['src'], $img_attributes['srcset'] ) ) {
+			unset( $img_attributes['src'] );
+		}
+
+		// Add media query if it's going to be something other than just `min-width: 0px`.
+		if ( $minimum_viewport_widths[ $i ] > 0 || isset( $minimum_viewport_widths[ $i + 1 ] ) ) {
+			$media_query = sprintf( '( min-width: %dpx )', $minimum_viewport_widths[ $i ] );
+			if ( isset( $minimum_viewport_widths[ $i + 1 ] ) ) {
+				$media_query .= sprintf( ' and ( max-width: %dpx )', $minimum_viewport_widths[ $i + 1 ] - 1 );
+			}
+			$img_attributes['media'] = $media_query;
+		}
+
+		// Construct preload link.
+		$link_tag = '<link data-ilo-added-tag rel="preload" fetchpriority="high" as="image"';
+		foreach ( array_filter( $img_attributes ) as $name => $value ) {
+			// Map img attribute name to link attribute name.
+			if ( 'srcset' === $name || 'sizes' === $name ) {
+				$name = 'image' . $name;
+			} elseif ( 'src' === $name ) {
+				$name = 'href';
+			}
+
+			$link_tag .= sprintf( ' %s="%s"', $name, esc_attr( $value ) );
+		}
+		$link_tag .= ">\n";
+
+		$preload_links[] = $link_tag;
+	}
+
+	return implode( '', $preload_links );
+}
+
+/**
+ * Optimizes template output buffer.
+ *
+ * @since n.e.x.t
+ * @access private
+ *
+ * @param string $buffer Template output buffer.
+ * @return string Filtered template output buffer.
+ */
+function ilo_optimize_template_output_buffer( string $buffer ): string {
+	$slug = ilo_get_url_metrics_slug( ilo_get_normalized_query_vars() );
+	$post = ilo_get_url_metrics_post( $slug );
+
+	// No URL metrics are present, so there's nothing we can do.
+	if ( ! $post ) {
+		return $buffer;
+	}
+
+	$url_metrics = ilo_parse_stored_url_metrics( $post );
+
+	$breakpoint_max_widths                   = ilo_get_breakpoint_max_widths();
+	$url_metrics_grouped_by_breakpoint       = ilo_group_url_metrics_by_breakpoint( $url_metrics, $breakpoint_max_widths );
+	$lcp_elements_by_minimum_viewport_widths = ilo_get_lcp_elements_by_minimum_viewport_widths( $url_metrics_grouped_by_breakpoint );
+	$all_breakpoints_have_url_metrics        = count( array_filter( $url_metrics_grouped_by_breakpoint ) ) === count( $breakpoint_max_widths ) + 1;
+
+	// TODO: Handle case when the LCP element is not an image at all, but rather a background-image.
+	// Prepare to set fetchpriority attribute on the image when all breakpoints have the same LCP element.
+	if (
+		// All breakpoints share the same LCP element (or all have none at all).
+		1 === count( $lcp_elements_by_minimum_viewport_widths )
+		&&
+		// The breakpoints don't share a common lack of an LCP element.
```

Comment:
Isn't this rather "lack of an LCP image"?

Reply:
At the moment, yes. But other LCP elements will be supported, namely elements with background images. There is a TODO for this above:

https://github.com/WordPress/performance/blob/83d0362170501b5356cea8cb403ba29913ea0ce6/modules/images/image-loading-optimization/optimization.php#L108

Reply:
Nit-pick, but could you still adjust this comment to say "LCP image"? Regardless of potential background image support in the future, my point here is mostly that "lack of an LCP element" is not a thing, as every page has an LCP element.

Reply:
Done in 185ebf6c